### PR TITLE
Improve Swift build performance by avoiding unnecessary recompilation of Swift modules

### PR DESCRIPTION
**Pull request by xianwen**
## TL;DR;
Update `SwiftCompile` to support `SupportsInputBasedRuleKey` interface, so that we could avoid unnecessary Swift module recompilation if its input Swift modules' `.swiftmodule` artifacts weren't changed.

## Details
Suppose Swift moduleA depends on Swift moduleB.

The current behavior of Buck is that whenever you touch the source code of moduleB, moduleA will always be rebuilt, even if you didn't change the public interface of moduleB. This has a heavy cost and unnecessarily slows incremental builds a lot.

This PR avoids the recompilation of moduleA in this given example by evaluating the md5sum of `moduleB.swiftmodule`, which is a binary superset of moduleB's all public interfaces. As long as a Swift module's `.swiftmodule` remains unchanged after its compilation, it is safe to not recompile its dependents.

I verified this PR is working in an Airbnb demo project Buck Sample here:
https://github.com/airbnb/BuckSample/pull/132

By applying this PR internally at the Airbnb's fork of Buck here, https://github.com/airbnb/buck/pull/23/, our p50 of local iOS build time has reduced more than 30% in the last 2 weeks.

### CODE REVIEW

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 1]
```diff
@@ -475,4 +477,18 @@ public SourcePath getObjCGeneratedHeaderPath() {
...
+   * swiftmodule file contains the public interface for a module, and is basically a binary file
+   * format equivalent to header files for a C framework or library.
+   *
+   * A swiftmodule file contains serialized ASTs (and possibly SIL), it conforms to
+   * Swift Binary Serialization Format, more details about this binary format can be found here:
+   * https://github.com/apple/swift/blob/7e6d62dae4bae4eb3737a6f76c0e51534c1bcca3/docs/Serialization.rst.
+   */
+  public SourcePath getSwiftModuleOutputPath() {
+    return ExplicitBuildTargetSourcePath.of(getBuildTarget(),
+      Paths.get(outputPath.toString(), moduleName + ".swiftmodule"));
```
Comment (milend) [inline src/com/facebook/buck/swift/SwiftCompile.java:492]: This can produce a different module name than what's actually produced by the compilation in the case of sanitisation. Instead, it should just return the precomputed `modulePath` like so:

    return ExplicitBuildTargetSourcePath.of(getBuildTarget(), modulePath);

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 2]
```diff
@@ -65,13 +66,14 @@
...
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.Optional;
 import java.util.SortedSet;
 import java.util.function.Function;
 
 /** A build rule which compiles one or more Swift sources into a Swift module. */
-public class SwiftCompile extends AbstractBuildRule {
+public class SwiftCompile extends AbstractBuildRule implements SupportsInputBasedRuleKey {
```
Comment (milend) [inline src/com/facebook/buck/swift/SwiftCompile.java:76]: This change can potentially break things because there's still a bunch of fields which are not part of the rulekey. This becomes problematic if any of those unannotated keys contains data computed by some transitive dependency and that data is part of the transitive rule's rulekey.

I'd need to audit all of the remaining fields to ensure that this is the case but I haven't had the time to do that yet.

On the correctness note, a quick look reveals that some stuff is already broken: e.g., `enableObjcInterop` should have an `@AddToRuleKey` annotation.

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 1 above)
Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:492]: 👍

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 2 above)
Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:76]: I see, thx for the review.

Would it be enough for us to review all the "private final" fields in `SwiftCompile` class? Or do we need to review the fields in its super class as well?

If former, then the fields we shall review would be the following:
```
  private final Path objectFilePath;
  private final Path modulePath;
  private final Path moduleObjectPath;
  private final ImmutableList<Path> objectPaths;
  private final Optional<AbsPath> swiftFileListPath;
  private final Path headerPath;
  private final CxxPlatform cxxPlatform;
  private final ImmutableSet<FrameworkPath> frameworks;
  private final boolean enableObjcInterop;
  private final SwiftBuckConfig swiftBuckConfig;
  private BuildableSupport.DepsSupplier depsSupplier;
```

Which means in order to safely apply this change, **we should review these fields and make sure all of them whenever necessary should have that `@AddToRuleKey` annotation.**

Is my understanding above correct? Thanks!

File: src/com/facebook/buck/cxx/FrameworkPathArg.java [hunk 3]
```diff
@@ -23,8 +23,8 @@
 import java.util.Objects;
 
 /** A base class for {@link Arg}s which wrap a {@link FrameworkPath}. */
-abstract class FrameworkPathArg implements Arg {
-  @AddToRuleKey final ImmutableSortedSet<FrameworkPath> frameworkPaths;
+public abstract class FrameworkPathArg implements Arg {
```
Comment (xianwen) [inline src/com/facebook/buck/cxx/FrameworkPathArg.java:26]: Made it `public` so that `com/facebook/buck/swift/SwiftLibraryDescription` can create a subclass of it.

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 4]
```diff
@@ -82,29 +82,36 @@
...
+  @AddToRuleKey(stringify = true)
   private final Path modulePath;
+  @AddToRuleKey(stringify = true)
   private final Path moduleObjectPath;
+  @AddToRuleKey(stringify = true)
   private final ImmutableList<Path> objectPaths;
   private final Optional<AbsPath> swiftFileListPath;
 
   @AddToRuleKey private final boolean shouldEmitSwiftdocs;
+  @AddToRuleKey private final boolean useModulewrap;
```
Comment (xianwen) [inline src/com/facebook/buck/swift/SwiftCompile.java:96]: From `swiftBuckConfig`

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 5]
```diff
@@ -82,29 +82,36 @@
...
+  @AddToRuleKey private final boolean compileForceCache;
```
Comment (xianwen) [inline src/com/facebook/buck/swift/SwiftCompile.java:97]: From `swiftBuckConfig`

File: src/com/facebook/buck/swift/SwiftLibraryDescription.java [hunk 6]
```diff
@@ -287,8 +292,15 @@ public BuildRule createBuildRule(
           cxxPlatform.getCpp().resolve(graphBuilder, buildTarget.getTargetConfiguration());
 
       BuildTarget buildTargetCopy = buildTarget;
+
+      ImmutableSortedSet.Builder<FrameworkPath> builder = ImmutableSortedSet.naturalOrder();
```
Comment (xianwen) [inline src/com/facebook/buck/swift/SwiftLibraryDescription.java:296]: This is duplicated code that should be refactored into a method if we want to move forward with this design.
I'm not doing it now because I prefer to adopt a different design.

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 7]
```diff
@@ -203,15 +213,7 @@ private SwiftCompileStep makeCompileStep(SourcePathResolverAdapter resolver) {
...
-    Function<FrameworkPath, Path> frameworkPathToSearchPath =
-        CxxDescriptionEnhancer.frameworkPathToSearchPath(cxxPlatform, resolver);
-
-    compilerCommand.addAll(
-        Streams.concat(frameworks.stream(), cxxDeps.getFrameworkPaths().stream())
-            .filter(x -> !x.isSDKROOTFrameworkPath())
-            .map(frameworkPathToSearchPath)
-            .flatMap(searchPath -> ImmutableSet.of("-F", searchPath.toString()).stream())
-            .iterator());
+    frameworksArg.appendToCommandLine(compilerCommand::add, resolver);
```
Comment (xianwen) [inline src/com/facebook/buck/swift/SwiftCompile.java:216]: @milend , I'm not sure if I understand you correctly regarding using a `frameworksArg` to replace both `cxxPlatform` and `frameworks`.

This usage of `frameworksArg.appendToCommandLine` isn't quite consistent with the other `compilerCommand.addAll` calls nearby. I would prefer to still use the same old `compilerCommand.addAll` here for this `frameworksArg` as well.

My proposal is:
Instead of passing in one single `frameworksArg`, we pass in `Function<FrameworkPath, Path> frameworkPathToSearchPath` and `ImmutableSortedSet<FrameworkPath> frameworkPaths` into `SwiftCompile`, and mark both as `@AddToRuleKey`.

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 8]
```diff
@@ -402,7 +404,7 @@ public SourcePath getSourcePathToOutput() {
             .toToolFlags(
                 resolver,
                 PathShortener.byRelativizingToWorkingDir(getProjectFilesystem().getRootPath()),
-                CxxDescriptionEnhancer.frameworkPathToSearchPath(cxxPlatform, resolver),
+                frameworkPathToSearchPath,
```
Comment (xianwen) [inline src/com/facebook/buck/swift/SwiftCompile.java:407]: According to the comment above:
>  Arg list can't simply be passed in since the current implementation of toToolFlags drops the
> dependency information.

I don't like the idea of making `frameworkArgs.frameworkPathToSearchPath` public, so I chose to pass it in directly instead and marked it as `@AddToRuleKey`.

However, this diminishes the benefit of having `frameworkArgs`. And it's the other reason why I think it might make sense to not use that new `frameworkArgs`, but to use `Function<FrameworkPath, Path> frameworkPathToSearchPath` and `ImmutableSortedSet<FrameworkPath> frameworkPaths` directly instead.

Comment (milend) [reply src/com/facebook/buck/swift/SwiftCompile.java:407]: @xianwen:
> I don't like the idea of making frameworkArgs.frameworkPathToSearchPath public, so I chose to pass it in directly instead and marked it as @AddToRuleKey

Agreed, it should not be exposed.

> Arg list can't simply be passed in since the current implementation of toToolFlags drops the dependency information.

I have a suspicion that this comment might be out of date (it was added back in 2017) and `CxxToolFlags` is annotated with `AddsToRuleKey`.

If we wanted to do this properly, we would pass `argsFromDeps` to the constructor, together with a single `Arg` for the frameworks. But it's safer to keep the current behaviour as changing it would require further validation.

File: src/com/facebook/buck/swift/SwiftCompile.java (hunk 7 above)
Comment (milend) [reply src/com/facebook/buck/swift/SwiftCompile.java:216]: @xianwen:

> Instead of passing in one single frameworksArg, we pass in Function<FrameworkPath, Path> frameworkPathToSearchPath and ImmutableSortedSet<FrameworkPath> frameworkPaths into SwiftCompile, and mark both as @AddToRuleKey.

The type of `frameworkPathToSearchFunction` should be `AddsToRuleKeyFunction<FrameworkPath, Path>`, agreed about the rest.

We basically have to go down that route because of the usage of `frameworkPathToSearchPath` in the `toToolFlags()` call.

File: src/com/facebook/buck/swift/SwiftLibraryDescription.java [hunk 9]
```diff
@@ -530,4 +554,32 @@ public static boolean isSwiftTarget(BuildTarget buildTarget) {
      */
     Optional<String> getTargetSdkVersion();
   }
+
+  private static class FrameworkLinkerArgs extends FrameworkPathArg {
+    @AddToRuleKey final AddsToRuleKeyFunction<FrameworkPath, Path> frameworkPathToSearchPath;
+
+    public FrameworkLinkerArgs(
```
Comment (milend) [inline src/com/facebook/buck/swift/SwiftLibraryDescription.java:561]: @xianwen: We do not need this anymore if we just pass the list of frameworks + function.

**Comment by milend** [comment]
@xianwen: Once you address the feedback and take the revised approach, can you please squash all changes into a single commit?

File: src/com/facebook/buck/swift/SwiftLibraryDescription.java (hunk 9 above)
Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftLibraryDescription.java:561]: No, we don't.
I'll remove this and also update a few other places accordingly.

**Comment by xianwen** [comment]
> @xianwen: Once you address the feedback and take the revised approach, can you please squash all changes into a single commit?

Of course, will do.

**Comment by xianwen** [comment]
@milend, all code review feedbacks are addressed and all commits are squashed into one, PTAL at your convenience, thanks!

**Comment by milend** [comment]
@xianwen: all looks good, will merge.

File: src/com/facebook/buck/swift/SwiftCompile.java [hunk 10]
```diff
@@ -82,29 +83,36 @@
...
   @AddToRuleKey private final Optional<String> version;
   @AddToRuleKey private final ImmutableList<? extends Arg> compilerFlags;
 
+  @AddToRuleKey(stringify = true)
   private final Path headerPath;
-  private final CxxPlatform cxxPlatform;
-  private final ImmutableSet<FrameworkPath> frameworks;
+  @AddToRuleKey private final ImmutableSet<FrameworkPath> frameworks;
+  @AddToRuleKey private final AddsToRuleKeyFunction<FrameworkPath, Path> frameworkPathToSearchPath;
+  @AddToRuleKey private final Flavor flavor;
```
Comment (milend) [inline src/com/facebook/buck/swift/SwiftCompile.java:111]: @xianwen: This requires `@AddToRuleKey(stringify = true)` as `Flavor` is not a supported type for rulekey computation.

Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:111]: Oops, sorry about that, I'll send you another PR fixing it up.

Comment (xianwen) [reply src/com/facebook/buck/swift/SwiftCompile.java:111]: https://github.com/facebook/buck/pull/2475

Comment (milend) [reply src/com/facebook/buck/swift/SwiftCompile.java:111]: Merged, thanks.
